src/unix/signal.c

```diff
@@ -693,45 +668,38 @@ static void setup_sigframe(thread t, int signum, struct siginfo *si, void * ucon
      * to be carried over */
     runtime_memcpy(t->sigframe, t->frame, sizeof(u64) * FRAME_MAX);
 
-    /* return to signal trampoline */
-    t->sigframe[FRAME_RIP] = u64_from_pointer(signal_trampoline);
-
     sig_debug("sa->sa_flags 0x%lx\n", sa->sa_flags);
 
     /* check for altstack */
     if (sa->sa_flags & SA_ONSTACK) {
         t->sigframe[FRAME_RSP] = 0; /* TODO */
         halt("SA_ONSTACK ...\n");
-    } else {
-        /* 16-byte alignment; avoid redzone */
-        t->sigframe[FRAME_RSP] = (t->sigframe[FRAME_RSP] & ~15) - 128;
     }
 
-    /* arguments to trampoline */
-    t->sigframe[FRAME_RDI] = signum;
-    t->sigframe[FRAME_RSI] = u64_from_pointer(sa->sa_handler);
+    /* 16-byte alignment; avoid redzone */
+    t->sigframe[FRAME_RSP] = (t->sigframe[FRAME_RSP] & ~15) - 128;
+
+    /* create space for rt_sigframe */
+    t->sigframe[FRAME_RSP] -= pad(sizeof(struct rt_sigframe), 16);
+
+    /* setup sigframe for user sig trampoline */
+    struct rt_sigframe *frame = (struct rt_sigframe *)t->sigframe[FRAME_RSP];
+    frame->pretcode = sa->sa_restorer;
 
-    /* copy siginfo + ucontext onto the stack */
     if (sa->sa_flags & SA_SIGINFO) {
-        struct ucontext * dest_uctx;
-        siginfo_t * dest_si;
-
-        t->sigframe[FRAME_RSP] -= pad(sizeof(struct siginfo), 16);
-        dest_si = pointer_from_u64(t->sigframe[FRAME_RSP]);
-        sig_debug("copying siginfo to [%p, %p)\n", dest_si, ((void *)dest_si) + sizeof(struct siginfo));
-        runtime_memcpy(dest_si, si, sizeof(struct siginfo));
-        t->sigframe[FRAME_RDX] = t->sigframe[FRAME_RSP];
-
-        t->sigframe[FRAME_RSP] -= pad(sizeof(struct ucontext), 16);
-        dest_uctx = pointer_from_u64(t->sigframe[FRAME_RSP]);
-        sig_debug("copying ucontext to [%p, %p)\n", dest_uctx, ((void *)dest_uctx) + sizeof(struct ucontext));
-        setup_ucontext(dest_uctx, sa, si, t->frame);
-        t->sigframe[FRAME_R8] = t->sigframe[FRAME_RSP];
+        runtime_memcpy(&(frame->info), si, sizeof(struct siginfo));
+        setup_ucontext(&(frame->uc), sa, si, t->frame);
+        t->sigframe[FRAME_RSI] = (u64)&(frame->info);
```

Comment:
I guess for the sake of consistency you could replace with u64_from_pointer() ...